runtime/lua/vim/lsp.lua

```diff
@@ -0,0 +1,791 @@
+    filetypes = config.filetype
+  else
+    error("config.filetype must be a string or a list of strings")
+  end
+
+  local offset_encoding = config.offset_encoding and validate_encoding(config.offset_encoding) or VALID_ENCODINGS.UTF16
+
+  local cmd, cmd_args
+  if type(config.cmd) == 'string' then
+    -- Use a shell to execute the command if it is a string.
+    cmd = vim.api.nvim_get_option('shell')
+    cmd_args = {vim.api.nvim_get_option('shellcmdflag'), config.cmd}
+  elseif vim.tbl_islist(config.cmd) then
+    cmd = config.cmd[1]
+    cmd_args = {}
+    -- Don't mutate our input.
+    for i, v in ipairs(config.cmd) do
+      assert(type(v) == 'string', "config.cmd arguments must be strings")
+      if i > 1 then
+        table.insert(cmd_args, v)
+      end
+    end
+  else
+    error("cmd type must be string or list.")
+  end
+
+  LSP_CONFIGS[config.name] = {
+    user_config = config;
+    name = config.name;
+    offset_encoding = offset_encoding;
+    filetypes = filetypes;
+    cmd = cmd;
+    cmd_args = cmd_args;
+    cmd_env = config.cmd_env;
+    cmd_cwd = config.cmd_cwd;
+    capabilities = capabilities;
+    init_options = config.init_options;
+    on_init = config.on_init;
+  }
+
+  nvim_command(string.format(
+    "autocmd FileType %s ++once silent lua vim.lsp._start_client_by_name(%q)",
+    table.concat(filetypes, ','),
+    config.name))
+end
+
+function lsp._start_client_by_name(name)
+  local config = LSP_CONFIGS[name]
+  -- If it exists and is running, don't make it again.
+  if config.client_id and LSP_CLIENTS[config.client_id] then
+    -- TODO log here?
+    return
+  end
+  config.client_id = lsp.start_client(config)
+  vim.lsp.attach_to_buffer(0, config.client_id)
+
+  nvim_command(string.format(
+    "autocmd FileType %s silent lua vim.lsp.attach_to_buffer(0, %d)",
+    table.concat(config.filetypes, ','),
+    config.client_id))
+  return config.client_id
+end
+
+nvim_command("autocmd VimLeavePre * lua vim.lsp.stop_all_clients()")
+
+function lsp.get_client_by_id(client_id)
+  return LSP_CLIENTS[client_id]
+end
+
+function lsp.get_client_by_name(name)
+  local config = LSP_CONFIGS[name]
+  if config.client_id then
+    return LSP_CLIENTS[config.client_id]
+  end
+end
+
+function lsp.stop_client(client_id, force)
+  local client = LSP_CLIENTS[client_id]
+  if client then
+    client.stop(force)
+  end
+end
+
+function lsp.stop_all_clients(force)
+  for client_id, client in pairs(LSP_CLIENTS) do
+    client.stop(force)
+  end
+end
+
+--- Send a request to a server and return the response
+-- @param method [string]: Name of the request method
+-- @param params [table] (optional): Arguments to send to the server
+-- @param bufnr [number] (optional): The number of the buffer
+-- @param filetype [string] (optional): The filetype associated with the server
+-- @param server_name [string] (optional)
+--
+-- @returns: success?, request_id, cancel_fn
+function lsp.buf_request(bufnr, method, params, callback)
```

Comment:
I'm not sure, but I thought it would make it very clear what it was doing. It is operating on the buffer, and it kind of follows other naming schemes for neovim functions.